cores/arduino/wiring.c

```diff
@@ -22,44 +18,126 @@
 
 #include "wiring_private.h"
 
-// the prescaler is set so that timer0 ticks every 64 clock cycles, and the
+// CHANGE by Gijs Withagen, 18-02-2019
+// Prescaler can be changed in runtime and still have a correct millis()
+// the micros() just has a glitch, but counted form changing the prescaler
+// it gives proper relative values.
+//
+// Defalult the prescaler is set so that timer0 ticks every 64 clock cycles, and the
 // the overflow handler is called every 256 ticks.
-#define MICROSECONDS_PER_TIMER0_OVERFLOW (clockCyclesToMicroseconds(64 * 256))
+#define MICROSECONDS_PER_TIMER0_OVERFLOW (clockCyclesToMicroseconds(getPrescaler() * 256))
 
 // the whole number of milliseconds per timer0 overflow
 #define MILLIS_INC (MICROSECONDS_PER_TIMER0_OVERFLOW / 1000)
 
-// the fractional number of milliseconds per timer0 overflow. we shift right
-// by three to fit these numbers into a byte. (for the clock speeds we care
-// about - 8 and 16 MHz - this doesn't lose precision.)
-#define FRACT_INC ((MICROSECONDS_PER_TIMER0_OVERFLOW % 1000) >> 3)
-#define FRACT_MAX (1000 >> 3)
+// the fractional number of milliseconds per timer0 overflow.
+#define FRACT_INC ((MICROSECONDS_PER_TIMER0_OVERFLOW % 1000))
+#define FRACT_MAX (1000)
 
 volatile unsigned long timer0_overflow_count = 0;
-volatile unsigned long timer0_millis = 0;
-static unsigned char timer0_fract = 0;
+
+volatile unsigned long correction_millis = 0;
+volatile unsigned long correction_frac = 0;
+
+volatile unsigned long millis_correction_inc = 0;
+volatile unsigned long fract_correction_inc = 0;
+volatile unsigned int previous_prescaler = 0;
+
+// Returns the actual value of the prescaler of Timer 0 needed to calculate
+// the elapsed microseconds and milliseconds based on the timer0_overflow_count
+unsigned int getPrescaler() {
```

Comment:
Since this is about Timer0, maybe naming it `getPrescaler0` would be a good idea.